### PR TITLE
Add IAM fallback for ACM certificate import

### PR DESCRIPTION
**Pull request by nhakmiller**
## Background

@gavinelder reported that the deploy process fails when deployed in either of the `cn` regions at the upload ACM certificate step. Further investigation by @austinbyers showed that ACM is in fact not supported in those regions.

This change modifies the deploy process to fallback to uploading a certificate to IAM in cases where the ACM service is not available.

Closes: #67 

## Changes

- Added a check to see if the ACM service is available
- When the ACM service is not available, upload the self signed certificate to IAM instead of ACM

## Testing

- None yet, looking into setting up an account in `cn`

### CODE REVIEW

File: tools/mage/deploy_certificate.go [hunk 1]
```diff
@@ -90,6 +93,20 @@ func uploadLocalCertificate(awsSession *session.Session) (string, error) {
...
 	}
 
+	// Check if the ACM service is supported before tossing the private key out into the ether
+	acmClient := acm.New(awsSession)
+	_, err = acmClient.ListCertificates(&acm.ListCertificatesInput{MaxItems: aws.Int64(0)})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() == "SubscriptionRequiredException" {
+				// ACM is not supported in this region or for this user, fall back to IAM
+				fmt.Println("deploy: ACM not supported, falling back to IAM for certificate management")
```
Comment (austinbyers) [inline tools/mage/deploy_certificate.go]: Should we always use IAM? What's the difference?

**Comment by gavinelder** [comment]
I will look at testing this, in order to get a AWS-CN account you require a Chinese residence & if it's a corporation a CN National.

File: tools/mage/deploy_certificate.go (hunk 1 above)
Comment (nhakmiller) [reply tools/mage/deploy_certificate.go]: AWS recommends always using ACM for certificate management when it's available. That's the main reason we defaulted to it. ACM offers many more certificate management features like automatic renewal etc. but users most likely won't be taking advantage of those features for the self-signed certificates we upload for them.

The only other reason I would prefer to use ACM by default is that you cannot manage certificates imported via IAM through the AWS console. So if user's decide to setup their own certificate, they will have to use the CLI/SDK to delete the self signed certificate we uploaded. Of course they don't need to delete the certificate we uploaded if they don't want to, they can replace it with their own ACM/IAM certificate and just let the self signed IAM certificate we created sit there unused.

With that in mind, would you prefer to axe ACM altogether?

**Comment by nhakmiller** [comment]
> I will look at testing this, in order to get a AWS-CN account you require a Chinese residence & if it's a corporation a CN National.

Ah well that will be difficult for me to do as I have neither a Chinese residence nor am I a CN National. In this case I will leave testing of the issue to you (or any other users active in the CN partition), and feel free to post here with any issues. We're also on [gitter](https://gitter.im/runpanther/community) for IM.

**Comment by gavinelder** [comment]
@nhakmiller  This may need some more work,  


```
deploy: uploading load balancer certificate
deploy: WARNING no ACM certificate ARN provided and no certificate file provided, generating a self-signed certificate
Error: InvalidParameter: 1 validation error(s) found.
- minimum field value of 1, ListCertificatesInput.MaxItems.
```

**Comment by nhakmiller** [comment]
> deploy: uploading load balancer certificate

Interesting, the aws CLI has no issues with that parameter, I suppose the go SDK has more strict requirements than the python SDK. I've just upped that to a minimum of 1, although we may just remove ACM entirely and use IAM by default.

**Comment by gavinelder** [comment]
@nhakmiller it's passed now . 

```deploy: uploading load balancer certificate
deploy: ACM not supported, falling back to IAM for certificate management
deploy: panther-app: CreateChangeSet: CREATE_PENDING
deploy: panther-app: CreateChangeSet: CREATE_COMPLETE
deploy: panther-app: ExecuteChangeSet: CREATE_IN_PROGRESS
deploy: panther-app: ExecuteChangeSet: ROLLBACK_IN_PROGRESS```

**Comment by gavinelder** [comment]
```
deploy: ACM not supported, falling back to IAM for certificate management
Error: EntityAlreadyExists: The Server Certificate with name PantherCertificate already exists.
	status code: 409, request id: 4d1021db-6e2a-4d22-9982-5353c7ec30ae
```

**Comment by nhakmiller** [comment]
> ```
> deploy: ACM not supported, falling back to IAM for certificate management
> Error: EntityAlreadyExists: The Server Certificate with name PantherCertificate already exists.
> 	status code: 409, request id: 4d1021db-6e2a-4d22-9982-5353c7ec30ae
> ```

Thanks for your help and patience. I'm going to assume that this occurred when you tried to deploy Panther in the same account in a different region, or you tried re-deploying Panther before it finished?

In either case, this latest version should fix that particular issue I believe. I apologize for sending you so much untested code, I will spend some time testing the IAM functionality separate from the CN partition tomorrow.

**Comment by gavinelder** [comment]
The failure was linked to invalid service principles in the cloud formations I am currently working through the errors 1 by 1 as the documentation does not marry up 100% with implementation in CN

**Comment by nhakmiller** [comment]
> The failure was linked to invalid service principles in the cloud formations I am currently working through the errors 1 by 1 as the documentation does not marry up 100% with implementation in CN

I see. Is it safe to say at this point you are experiencing no issues with certificates during deployment (granted that the deployment isn't finishing for other reasons)? If so, I will probably merge this.

Additionally, I'd recommend you run `aws iam list-server-certificates` from the aws-cli, and delete old certificates with `aws iam delete-server-certificate` after you get it to work as every failed deploy is creating a new certificate (because the deploy is failing, so it is unable to find the old one even though it exists).

**Comment by nhakmiller** [comment]
I'm going to merge this now, as it sounds like any persisting deployment issues are not related to this certificate generation.

**Pull request merged**